### PR TITLE
Add support for ffmpeg 3.0.

### LINKED ISSUE #6310: ffmpeg 3.0 support
Add support for linking opencv with ffmpeg 3.0.

### PR DESCRIPTION
**Pull request by peters**
resolves #6310 

/cc @alalek

### CODE REVIEW

**Comment by alalek** [comment]
Thanks!
There are should be 3 ""#if"s for `LIBAVUTILS_BUILD` (1st,3rd,4th blocks) and one for `LIBAVCODEC_BUILD` (2nd block) following original comments.

File: modules/videoio/src/cap_ffmpeg_impl.hpp [hunk 1]
```diff
@@ -982,9 +1020,9 @@ bool CvCapture_FFMPEG::retrieveFrame(int, unsigned char** data, int* step, int*
         int aligns[AV_NUM_DATA_POINTERS];
         avcodec_align_dimensions2(video_st->codec, &buffer_width, &buffer_height, aligns);
         rgb_picture.data[0] = (uint8_t*)realloc(rgb_picture.data[0],
-                avpicture_get_size( AV_PIX_FMT_BGR24,
+                _opencv_ffmpeg_avimage_get_size( AV_PIX_FMT_BGR24,
```
Comment (alalek) [inline modules/videoio/src/cap_ffmpeg_impl.hpp]: `_opencv_ffmpeg_av_image_get_buffer_size`?

Comment (peters) [inline modules/videoio/src/cap_ffmpeg_impl.hpp]: Gawd, must have messed up when squashing. I'll correct it.

**Comment by peters** [comment]
The ifdef's has been updated.

File: modules/videoio/src/cap_ffmpeg_impl.hpp [hunk 2]
```diff
@@ -398,6 +403,39 @@ inline int _opencv_ffmpeg_interrupt_callback(void *ptr)
...
+#if LIBAVCODEC_BUILD >= (LIBAVCODEC_VERSION_MICRO >= 100 \
+    ? CALC_FFMPEG_VERSION(55, 25, 100) : CALC_FFMPEG_VERSION(55, 16, 0))
+    av_packet_unref(pkt);
+#else
+    av_free_packet(pkt);
+#endif
+};
+
+static
+inline void _opencv_ffmpeg_av_image_fill_arrays(void *frame, const uint8_t *ptr, enum AVPixelFormat pix_fmt, int width, int height)
```
Comment (alalek) [inline modules/videoio/src/cap_ffmpeg_impl.hpp]: Please remove "const" here: `const uint8_t *ptr`

Comment (peters) [inline modules/videoio/src/cap_ffmpeg_impl.hpp]: :+1:

**Comment by alalek** [comment]
:+1:

**Pull request merged**